javascript/ql/test/experimental/Security/CWE-073/ExpressHbsLFR.ql

```diff
@@ -0,0 +1,44 @@
+/**
+ * @name Express-Hbs Local File Read and Potential RCE
+ * @description Writing user input directly to res.render of ExpressJS used with Hbs can lead to LFR
+ * @kind path-problem
+ * @problem.severity error
+ * @precision high
+ * @id js/express-hbs-lfr
+ * @tags security
+ *       external/cwe/cwe-073
+ *       external/cwe/cwe-094
+ */
+
+import javascript
+import DataFlow
+import PathGraph
+import Express
+import semmle.javascript.DynamicPropertyAccess
+
+predicate isUsingHbsEngine() {
+  exists(MethodCallExpr method |
+    method.getMethodName() = "set" and
+    Express::appCreation().flowsToExpr(method.getReceiver()) and
+    method.getArgument(1).getStringValue().matches("hbs")
+  )
+}
+
+class HbsLFRTaint extends TaintTracking::Configuration {
+  HbsLFRTaint() { this = "HbsLFRTaint" }
+
+  override predicate isSource(Node node) { node instanceof RemoteFlowSource }
+
+  override predicate isSink(Node node) {
+    exists(MethodCallExpr mc |
+      Express::isResponse(mc.getReceiver()) and
+      mc.getMethodName() = "render" and
+      node.asExpr() = mc.getArgument(1) and
+      isUsingHbsEngine()
+    )
+  }
+}
+
+from HbsLFRTaint cfg, Node source, Node sink
+where cfg.hasFlow(source, sink)
+select source, sink
```

Comment:
The `select` clause of a `@kind path-problem` query must follow a certain pattern: 

```suggestion
from HbsLFRTaint cfg, PathNode source, PathNode sink
where cfg.hasFlowPath(source, sink)
select sink.getNode(), source, sink, "Template object injection due to $@.", source.getNode(),
  "user-provided value"
```

If you follow that pattern then the results will look nice.  
You can for example look at this query run: https://lgtm.com/query/3955073410786994877/

Reply:
Yes, this looks much better and can create a flow path.